d2/src/main/java/com/linkedin/d2/balancer/simple/ClusterLoadBalancerSubscriber.java

```diff
@@ -43,13 +46,24 @@ public ClusterLoadBalancerSubscriber(SimpleLoadBalancerState simpleLoadBalancerS
   @Override
   protected void handlePut(final String listenTo, final ClusterProperties discoveryProperties)
   {
-    // TODO: pick stable or canary configs if canary exists
     if (discoveryProperties != null)
     {
+      ClusterProperties pickedProperties = discoveryProperties;
+      if (discoveryProperties instanceof ClusterPropertiesWithCanary) {
```

Comment:
Sorry I missed the original review where the canary classes were introduced, I should have brought this up there. What was the motivation for creating subclasses (`ClusterPropertiesWithCanary` and `ServicePropertiesWithCanary`) instead of adding the new data to the base classes and having it be null by default? The issue with using subclassing is it locks you in to a strict structure and doesn't allow for composition. As you mention below with the current model the failover data would need to extend the canary classes, but these features aren't necessarily related in that way. 
Is this necessary to preserve binary compatibility on the base classes? If not I think modifying them and dropping the subclasses would be a cleaner approach.

Reply:
added composite classes and tests for picking between canary and stable configs.